README.md
# LRcell tutorial

**Package version**: 0.99.5


## Introduction
The goal of LRcell is to identify specific sub-cell types that drives the changes 
observed in a bulk RNA-seq differential gene expression experiment. To achieve this, 
LRcell utilizes sets of cell marker genes acquired from single-cell RNA-sequencing (scRNA-seq) 
as indicators for various cell types in the tissue of interest. Next, for each cell type, 
using its marker genes as indicators, we apply Logistic Regression on the complete set of genes with 
differential expression p-values to calculate a cell-type significance p-value. 
Finally, these p-values are compared to predict which one(s) are likely to be responsible 
for the differential gene expression pattern observed in the bulk RNA-seq experiments. 
LRcell is inspired by the [LRpath](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2639007/) algorithm developed by Sartor et al., originally designed for pathway/gene set enrichment analysis. LRcell contains three major components: LRcell analysis, plot generation and marker gene selection. All modules in this package are written in R. 

**Pre-loaded marker genes**

LRcell provides marker genes in the Prefrontal Cortex (pFC) human brain region, human PBMC and 
nine mouse brain regions (Frontal Cortex, Cerebellum, Globus Pallidus, Hippocampus, Entopeduncular, 
Posterior Cortex, Striatum, Substantia Nigra and Thalamus). The human brain dataset describes 
single-cell gene expression profiles of control samples in [Major Depressive Disorder (MDD) disease studies](https://www.nature.com/articles/s41593-020-0621-y). The mouse brain dataset comes from 
a [whole mouse brain analysis](https://www.sciencedirect.com/science/article/pii/S0092867418309553) and the data is also available at [DropViz](http://dropviz.org/).

## Workflow

### Installation

This is a R Bioconductor package and it can be installed by using `BiocManager`.
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager") ## this will install the BiocManager package
BiocManager::install("LRcell")
```

To check whether LRcell package is successfully installed:
```{r}
library(LRcell)
```

### LRcell usage

Once we have LRcell package loaded, we can start using it to analyze the transcriptional
engagement of cell types or clusters. LRcell **takes both single-cell marker genes list 
and p-values of bulk DE genes as input** to calculate the enrichment of cell-type specific
marker genes in the ranked DE genes. 

As mentioned above, LRcell provides single-cell marker genes list in 1 human brain region
(Prefrontal Cortex), 1 human PBMC and 9 mouse brain regions (Frontal Cortex, Cerebellum, Globus Pallidus, 
Hippocampus, Entopeduncular, Posterior Cortex, Striatum, Substantia Nigra and Thalamus). 

- The human brain data comes from control samples in [Major Depressive Disorder studies](https://www.nature.com/articles/s41593-020-0621-y). 
- The human PBMC data comes from volunteers in [HIV vaccine trial](https://www.biorxiv.org/content/10.1101/2020.10.12.335331v1) at time point of immediately before.
- The mouse data comes from the [whole brain single-cell RNA-seq experiments](https://www.sciencedirect.com/science/article/pii/S0092867418309553). Another resource for this dataset is [DropViz](http://dropviz.org/).

The data is stored in another Bioconductor ExperimentHub package named [LRcellTypeMarkers](https://github.com/marvinquiet/LRcellTypeMarkers). Users can access the data through ExperimentHub by:

```{r}
## for installing ExperimentHub
BiocManager::install("ExperimentHub")

## query data
library(ExperimentHub)
eh <- ExperimentHub::ExperimentHub()
eh <- AnnotationHub::query(eh, "LRcellTypeMarkers")
eh  ## this will list out EH number to access the calculated gene enrichment scores

## get mouse brain Frontal Cortex enriched genes
enriched.g <- eh[["EH4548"]]
marker.g <- get_markergenes(enriched_genes, method="LR", topn=100)
```

Users are also encouraged to process a read count matrix with cell annotation information into a gene enrichment scores matrix.
```{r eval=FALSE}
enriched.g <- LRcell_gene_enriched_scores(expr, annot, power=1, parallel=TRUE, n.cores=4)
```
Here, `expr` is a read count matrix with rows as genes and columns as cells. `annot` is a named-vector with names as cell names (which is in accordance with the column names of `expr`) and values as annotated cell types. `power` is a hyper-parameter controlling how much penalty for the proportion of cells expressing certain gene. `parallel` and `n.cores` are two parameters for executing function in parallel to accelerate the calculation.

#### Directly indicate species and brain region in LRcell

Compared to processing data yourself, a much easier way is to indicate species and brain region or tissue. 
In this way, marker genes are extracted from ExperimentHub accordingly. For example, we can use mouse Frontal Cortex
marker genes to do LRcell analysis on the example bulk experiment. (The example contains 23, 420 genes along with p-values calculated from [DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html). 
Data is processed from a mouse Alzheimer's disease model (GEO: [GSE90693](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE90693)), 
which is 6 months after treatment in Frontal Cortex brain region.)

Here, we take method as Linear Regression as example:
```{r}
data("example_gene_pvals")
res <- LRcell(gene.p = example_gene_pvals,
              marker.g = NULL,
              species = "mouse",
              region = "FC",
              method = "LiR")
FC_res <- res$FC
# exclude leading genes for a better view
sub_FC_res <- subset(FC_res, select=-lead_genes)
head(sub_FC_res[order(sub_FC_res$p.value), ])
```

The top 6 significant lines of the result:

| ID                            | genes\_num | coef         | p\.value        | FDR             | cell\_type       |
|-------------------------------|------------|--------------|-----------------|-----------------|------------------|
| FC\_8\-2\.Astrocytes          | 100        | 0\.019701817 | 4\.761701e\-121 | 3\.856978e\-119 | Astrocytes       |
| FC\_11\-4\.unknown            | 90         | 0\.089774550 | 1\.327954e\-93  | 5\.378213e\-92  | unknown          |
| FC\_11\-1\.Microglia          | 98         | 0\.056507809 | 2\.718086e\-91  | 7\.338832e\-90  | Microglia        |
| FC\_9\-1\.Oligodendrocytes    | 100        | 0\.008969389 | 6\.032076e\-40  | 1\.221495e\-38  | Oligodendrocytes |
| FC\_9\-3\.Oligodendrocytes    | 98         | 0\.006969152 | 1\.660017e\-35  | 2\.689227e\-34  | Oligodendrocytes |
| FC\_9\-4\.Oligodendrocytes    | 97         | 0\.006447281 | 6\.246377e\-31  | 8\.432608e\-30  | Oligodendrocytes |


You can also try plotting out the result:
```{r} 
g <- plot_manhattan_enrich(FC_res, sig.cutoff = .05, label.topn = 5) ## a ggplot2 object is returned
g
```

## Calculate gene enrichment scores from expression dataframe
The gene enrichment scores calculation is based on algorithm in this [science paper](https://science.sciencemag.org/content/352/6291/1326.long)
which takes both enrichment of genes in certain cell types and fraction of cells expressing the gene into consideration. 

`LRcell_gene_enriched_scores()` function takes the gene-by-cell expression matrix and a cell-type annotation as input, which means a clustering algorithm and/or cell-type annotations should be done first. The columns of the expression matrix should match with cell names in the annotation vector. 

Below is a naive example on how to utilize this function to select marker genes.

**Example gene-by-cell expression matrix:**

| | cell1 | cell2 | cell3 | cell4 | cell5 | cell6 | cell7 | cell8| cell9 | cell10|
|---|---|---|---|---|---|---|---|---|---|---|
| gene1 | 3 | 0 | 2 | 8 | 10 | 6 | 1 | 0 | 0 | 2 |
| gene2 | 7 | 5 | 8 | 1 | 0 | 5 | 2 | 3 | 2 | 1 |
| gene3 | 8 | 10 | 6 | 7 | 8 | 9 | 5 | 8 | 6 | 8 |

**Example cell annotation**
| cell ID | cell1 |  cell2 | cell3 | cell4 | cell5 | cell6 | cell7 | cell8| cell9 | cell10|
|---|---|---|---|---|---|---|---|---|---|---|
| Annotation | celltype1 | celltype1 | celltype1 | celltype2 | celltype2 | celltype2 | celltype3 | celltype3 | celltype3 | celltype3 |

This toy example contains 3 genes and 10 cells. As you can tell from the matrix, __gene1__ is 
a marker gene of __celltype2__; __gene2__ is a marker gene of __celltype1__; __gene3__ is a house keeping gene.


```{r}
# generate a simulated gene*cell read counts matrix
n.row <- 3; n.col <- 10
sim.expr <- matrix(0, nrow=n.row, ncol=n.col)
rownames(sim.expr) <- paste0("gene", 1:n.row)
colnames(sim.expr) <- paste0("cell", 1:n.col)

# generate a simulated annotation for cells
sim.annot <- c(rep("celltype1", 3), rep("celltype2", 3), rep("celltype3", 4))
names(sim.annot) <- colnames(sim.expr)

sim.expr['gene1', ] <- c(3, 0, 2, 8, 10, 6, 1, 0, 0, 2) # marker gene for celltype2
sim.expr['gene2', ] <- c(7, 5, 8, 1, 0, 5, 2, 3, 2, 1) # marker gene for celltype1
sim.expr['gene3', ] <- c(8, 10, 6, 7, 8, 9, 5, 8, 6, 8) # house keeping

# generating the enrichment score 
enriched_res <- LRcell_gene_enriched_scores(expr = sim.expr,
                            annot = sim.annot, parallel = FALSE)
enriched_res
```

After using `LRcell_gene_enriched_scores()` function, we get the enriched scores for each gene in each cell type, which is:

| | celltype1 | celltype2 | celltype3|
|---|---|---|---|
| gene1 | 0.3472222 | 2.5 | 0.1171875 |
| gene2 | 1.960784 | 0.3921569 | 0.5882353 |
| gene3 | 1.066667 | 1.066667 | 0.9 |

For example, __gene2__ is a marker gene of __celltype1__, after calculation, it has the highest enrichment score in __celltype1__ among all three cell types.

Then we can choose marker genes using top 1 as threshold:

```{r}
marker_res <- get_markergenes(enriched.g = enriched_res,
                method = "LR", topn=1)
```

The result becomes:

| celltype | marker genes |
|---|---|
| celltype1 | gene2 |
| celltype2 | gene1 |
| celltype3 | gene3 |


NEWS.md
# LRcell 0.99.7
- Fix typo

# LRcell 0.99.6
- Edit Vignette

# LRcell 0.99.5
- Add PBMC dataset from ExperimentHub

# LRcell 0.99.4
- Debug test-LRcell.R and fix the problem due to the data uploaded
- Add PBMC datasets information in utils.R
- Edit description

# LRcell 0.99.3
- change both dependency back to R>=4.1
- R>=3.6 generates a warning
---

# LRcell 0.99.2
- change LRcellTypeMarkers dependency to R>=3.6
- change LRcell dependency to R>=3.6
---

# LRcell 0.99.1
- remove the LRcell.Rproj
- change the .gitignore file
---

# LRcell 0.99.0
- version 0.99.0 released
- Submitted to Bioconductor
---



tests/testthat.R
library(testthat)
library(LRcell)

test_check("LRcell")



tests/testthat/test-LRcell.R
context("Test LRcell")

test_that("LRcell works on provided example dataset (Alzheimer's disease)", {
    data(example_gene_pvals)
    res <- LRcell(example_gene_pvals, region="FC")
    expect_equal(res$FC[which.min(res$FC$p.value), ]$cell_type, "Microglia")
})

test_that("LRcell enriched gene generation", {
    # generate a simulated gene*cell read counts matrix
    n.row <- 3; n.col <- 10
    sim.expr <- matrix(0, nrow=n.row, ncol=n.col)
    rownames(sim.expr) <- paste0("gene", 1:n.row)
    colnames(sim.expr) <- paste0("cell", 1:n.col)

    # generate a simulated annotation for cells
    sim.annot <- c(rep("celltype1", 3), rep("celltype2", 3), rep("celltype3", 4))
    names(sim.annot) <- colnames(sim.expr)

    sim.expr['gene1', ] <- c(3, 0, 2, 8, 10, 6, 1, 0, 0, 2) # marker gene for celltype2
    sim.expr['gene2', ] <- c(7, 5, 8, 1, 0, 5, 2, 3, 2, 1) # marker gene for celltype1
    sim.expr['gene3', ] <- c(8, 10, 6, 7, 8, 9, 5, 8, 6, 8) # house keeping

    # manual calculation on the result should be
    manual.enriched.df <- data.frame("celltype1" = c(0.3472222, 1.960784, 1.066667),
                              "celltype2" = c(2.5, 0.3921569, 1.066667),
                              "celltype3" = c(0.1171875, 0.5882353, 0.9))
    rownames(manual.enriched.df) <- rownames(sim.expr)

    # code result
    code.enriched.df <- LRcell_gene_enriched_scores(sim.expr, sim.annot, parallel = FALSE)

    expect_equal(manual.enriched.df['gene2', 'celltype2'],
                 unlist(code.enriched.df['gene2', 'celltype2']),
                 tolerance = .01)
    expect_equal(manual.enriched.df['gene1', 'celltype3'],
                 unlist(code.enriched.df['gene1', 'celltype3']),
                 tolerance = .01)
})


R/data.R
#' Mapping between subclusters and cell types in Mouse Brain
#'
#' A named vector containing the subclusters as name and cell types as values
#' in Mouse Brain. The cell types are pre-annotated by the dataset, which
#' includes: Endothelial, FibroblastLike, Mural, Oligodendrocytes,
#' Polydendrocytes, Astrocytes and Microglia.
#'
#' @format A named vector with 565 subclusters:
#' \describe{
#'     Named vector with name as subclusters and values as cell types.
#' }
#' @source \url{http://dropviz.org/} under tab `data`
#' @usage data(mouse_celltypes)
#'
"mouse_celltypes"

#' Example gene_pvals named vector from mouse experiment.
#'
#' A named vector containing gene symbols as name and p-values as values. This
#' is from a mouse Alzheimer's disease model (GEO: GSE90693), specifically 6
#' months after treatment in Frontal Cortex brain region. In this dataset, we
#' expect to see the Microglia as the most enriched cell type.
#'
#' @format A named vector with 23,420 items
#'
#' @source  \url{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE90693}
#' `GSE90693_RawCountsData_TPR50_6months_AllRegions.txt.gz`
#' @usage data(example_gene_pvals)
#'
"example_gene_pvals"

#' Example marker genes from mouse FC brain region.
#'
#' A list of marker genes with names indicating cell types. We selected top 100
#' enriched genes from each subcluster as marker genes list.
#'
#' @format A named vector with 81 subclusters in mouse Frontal Cortex:
#' \describe{
#'     Named vector with name as subclusters and values as marker genes.
#' }
#' @source Caculated from gene enrichment scores
#'
#' @usage data(mouse_FC_marker_genes)
"mouse_FC_marker_genes"


#' An example output of LRcell using data \link{example_gene_pvals} and
#' \link{mouse_FC_marker_genes}.
#'
#' @format A data frame with 81 rows as mouse FC sub-clusters and 8 variables:
#' \describe{
#'     \item{ID}{The IDs of each marker genes, can be a cell type or cluster}
#'     \item{genes_num}{How many marker genes are contributing to the analysis}
#'     \item{coef}{The coefficients of Logistic Regression or Linear Regression}
#'     \item{odds_ratio}{The odds ratio quantifies association in Logistic
#'     Regression}
#'     \item{p.value}{The p-value calculated from the analysis}
#'     \item{FDR}{The FDR after BH correction}
#'     \item{lead_genes}{Genes that are contributing to the analysis}
#'     \item{cell_type}{Cell typ ename}
#' }
#'
#'@usage data(example_LRcell_res)
"example_LRcell_res"


R/plot.R
#' Manhattan plot for the enrichment of cell types
#'
#' This function draws out the LRcell result dataframe. In this function, we take
#' LRcell result dataframe and added cell types according to
#'
#' @name plot_manhattan_enrich
#'
#' @param lrcell_res LRcell result dataframe.
#'
#' @param sig.cutoff The p-value cutoff showing significance result of LRcell.
#'
#' @param label.topn A numeric number showing how many significant cell types
#' will be labeled.
#'
#' @return A ggplot2 object
#'
#' @importFrom dplyr group_by arrange summarize
#' @import ggplot2
#' @import ggrepel
#' @import utils
#' @importFrom magrittr %>%
#'
#' @export
#'
#' @examples
#' data(example_LRcell_res)
#' plot_manhattan_enrich(example_LRcell_res)
plot_manhattan_enrich <- function(lrcell_res,
                                sig.cutoff = 0.05,
                                label.topn = 5) {
    lr_tmp <- lrcell_res %>%
        dplyr::group_by(.data$cell_type) %>%
        dplyr::arrange(.data$cell_type)
    lr_tmp$pos <- as.numeric(rownames(lr_tmp))

    # axis center
    axisdf <- lr_tmp %>%
        dplyr::group_by(.data$cell_type) %>%
        dplyr::summarize(center = (max(.data$pos) + min(.data$pos)) / 2)

    thres_topn <- ifelse(nrow(lr_tmp) > label.topn, label.topn, nrow(lr_tmp))
    repel_thres <- sort(lr_tmp$FDR)[thres_topn]
    g <- ggplot2::ggplot(lr_tmp, aes(x = .data$pos, y = -log10(.data$FDR))) +
        # add points
        ggplot2::geom_point(aes(color = as.factor(.data$cell_type)), alpha = 0.8, size = 2) +
        ggplot2::geom_hline(yintercept = -log10(sig.cutoff), linetype = "dashed", color = "red") +
        # geom_label_repel(aes(label = ifelse(FDR <= repel_thres & FDR <= sig.cutoff, as.character(ID), "")),
        #   arrow = arrow(length = unit(0.03, "npc"), type = "closed", ends = "first"),
        #   nudge_y = 1
        # ) +
        ggrepel::geom_text_repel(aes(label = ifelse(.data$FDR <= repel_thres & .data$FDR <= sig.cutoff, as.character(.data$ID), "")),
                        force = 5) +
        ggplot2::scale_x_continuous(label = axisdf$cell_type, breaks = axisdf$center) +
        # add labs
        ggplot2::labs(x = "clusters") +
        ggplot2::theme_bw() +
        ggplot2::theme(
          axis.text.x = element_text(face = "bold", size = 12, angle = 45, hjust = 1),
          panel.grid.minor.x = element_blank()
        )
    g
}


#' Plot marker genes distribution on DE gene rank
#'
#' This function draws out the marker gene distribution for a certain cell type
#' (or cluster) on the DE gene rank list.
#'
#' @name plot_marker_dist
#'
#' @param markers Vector of marker genes from a cell type or cluster of interest.
#'
#' @param gene.p Named vector of gene-level pvalues from DEG analysis, i.e.
#' DESeq2, LIMMA
#'
#' @param colour Users can define the bar color they want on the ggplot2 object.
#'
#' @return A ggplot2 object
#'
#' @import ggplot2
#'
#' @export
#'
#' @examples
#' data(example_gene_pvals)
#' data(mouse_FC_marker_genes)
#' Oligos_markers <- mouse_FC_marker_genes[["FC_9-5.Oligodendrocytes_5"]]
#' plot_marker_dist(Oligos_markers, example_gene_pvals)
plot_marker_dist <- function(markers,
                            gene.p,
                            colour="red") {
    gene.p <- sort(gene.p)
    ticks <- match(markers, names(gene.p))
    ticks <- ticks[!is.na(ticks)]
    ticks_df <- data.frame(x=ticks)

    g <- ggplot2::ggplot() +
        ggplot2::geom_hline(yintercept = 0, colour="black") +
        ggplot2::geom_segment(data=ticks_df,
                              mapping=aes(x=.data$x, xend=.data$x, y=-0.5, yend=0.5),
                              size=0.3, colour = colour) +
        ggplot2::xlim(c(0, length(gene.p))) + ggplot2::ylim(c(-1, 1)) +
        ggplot2::labs(x = "Bulk DE Genes Ranking", y = "Indicator Bar Height",
             title="Marker Gene Distribution on DE genes") +
        ## add themes
        ggplot2::theme_bw() +
        ggplot2::theme(
            axis.text.x = element_text(face = "bold", size = 12, hjust = 1),
            panel.grid.minor.x = element_blank()
        )
    g
}




R/LRcell.R
#' Cell-type enrichment analysis for preranked gene set.
#'
#' This function wraps around \code{\link{LRcellCore}} in case of empty
#' inputs of the marker gene file and brain region.
#'
#' @name LRcell
#' @param gene.p Named vector of gene-level pvalues from DEG analysis, i.e.
#' DESeq2, LIMMA
#'
#' @param marker.g List of Cell-type specific marker genes derived from
#' single-cell RNA-seq. The name of the list is cell-type or cluster name, the
#' values are marker genes vectors or numeric named vectors. LRcell provides
#' marker genes list in different human/mouse brains, but users could use their
#' own marker gene list as input.
#' default: NULL
#'
#' @param species Either `mouse` or `human`, default: mouse.
#'
#' @param region Specific brain regions provided by LRcell. For mouse, LRcell
#' provides 9 brain regions: c("FC", "HC", "PC", "GP", "STR", "TH", "SN", "ENT",
#' "CB"). For human, LRcell provides c("pFC", "PBMC")
#'
#' @param method Either `logistic regression` or `linear regression`. Logistic
#' regression equally treats cell-type specific marker genes, however, if
#' certain values could determine the importance of marker genes, linear
#' regression can be performed, default: LR.
#'
#' @param min.size Minimal size of a marker gene set, will impact the balance of
#' labels
#'
#' @param sig.cutoff Cutoff for input genes pvalues, default: 0.05.
#'
#' @return A list with LRcell results. Each item represents a marker gene input.
#' Each item in this list is a statistics table. In the table, the row
#' represents the name of marker genes, and the columns are:
#'\itemize{
#'  \item ID The IDs of each marker genes, can be a cell type or cluster;
#'  \item genes_num How many marker genes are contributing to the analysis;
#'  \item coef The coefficients of Logistic Regression or Linear Regression;
#'  \item odds_ratio The odds ratio quantifies association in Logistic
#'  Regression;
#'  \item p-value The p-value calculated from the analysis;
#'  \item FDR The FDR after BH correction.
#'  \item lead_genes Genes that are contributing to the analysis;
#'}
#'
#' @import ExperimentHub
#' @import AnnotationHub
#'
#' @export
#' @examples
#' data(example_gene_pvals)
#' res <- LRcell(example_gene_pvals, species="mouse", region="FC", method="LR")
LRcell <- function(gene.p,
                marker.g = NULL,
                species = c("mouse", "human"),
                region = NULL,
                method = c("LR", "LiR"),
                min.size = 5,
                sig.cutoff = 0.05) {

    species <- match.arg(species)
    method <- match.arg(method)

    sc_marker_genes_list <- NULL
    regions <- NULL
    internal_data_ind <- TRUE # whether user uses internal data
    result_list <- NULL

    # user does not provide marker gene list
    if (is.null(marker.g)) {

        # user does not provide region information
        if (is.null(region)) {
            message("Because you did not choose a certain region, we will run all regions in this species for you.
It might take some time...")
            if (species == "mouse")
                regions <- MOUSE_REGIONS
            if (species == "human")
                regions <- HUMAN_REGIONS
        } else
            regions[1] <- region

        # get the list of marker genes
        for (reg in regions) {
            reg <- validate_region(species, reg)

            ## Use URL to acquire data, maybe needed in future if FTP server is ready
            # marker_genes_url <-
            #     "https://github.com/marvinquiet/LRcell/blob/master/marker_genes_lib/"
            # Github urls where enriched genes are stored
            # enriched_genes_url <- paste0(marker_genes_url,
            #                              paste0(species, '/', reg, 'enriched_genes.RDS'),
            #                              "?raw=true")
            #
            # N.TRIES <- 3  # try 3 times, refer to http://bioconductor.org/developers/how-to/web-query/
            # while (N.TRIES > 0L) {
            #     enriched_genes <- tryCatch(readRDS(url(enriched_genes_url)), error=identity)
            #     if (!inherits(enriched_genes, "error"))
            #         break
            #     N.TRIES <- N.TRIES - 1L
            # }
            #
            # if (N.TRIES == 0L) {
            #     stop("'getURL()' failed:",
            #          "\n  URL: ", enriched_genes_url,
            #          "\n  error: ", conditionMessage(enriched_genes),
            #          "\n Please check whether the Internect Connection is stable.")
            # }

            ## Use ExperimentHub to download data
            eh <- ExperimentHub::ExperimentHub()
            eh <- AnnotationHub::query(eh, "LRcellTypeMarkers")  ## query the data in AnnotationHub

            if (species == "mouse") {
                enriched_genes <- eh[[MOUSE_EXPHUB_MAPPING[reg]]]
            }

            if (species == "human") {
                enriched_genes <- eh[[HUMAN_EXPHUB_MAPPING[reg]]]
            }
            sc_marker_genes_list[[reg]] <- get_markergenes(enriched_genes, method)
        }
    } else {
        sc_marker_genes_list[["user"]] <- marker.g
        internal_data_ind <- FALSE
    }

    for (item in names(sc_marker_genes_list)) {
        marker_genes <- sc_marker_genes_list[[item]]
        res <- LRcellCore(gene.p = gene.p,
                        marker.g = marker_genes,
                        method = method,
                        min.size = min.size,
                        sig.cutoff = sig.cutoff)
        # if users use provided data, then add cell type to the dataframe, otherwise
        # LRcell uses ID as cell type
        if (internal_data_ind) {
            if (item %in% MOUSE_REGIONS) {
                ## set cell types for mouse brain data
                split_res <- strsplit(as.character(res$ID), "\\.")
                celltypes <- unlist(lapply(split_res, "[", 2))
                res$cell_type <- celltypes
            } else if ("pFC" == item) {
                ## set cell types for human pFC data
                split_res <- strsplit(as.character(res$ID), "_")
                celltypes <- unlist(lapply(split_res, "[", 1))
                res$cell_type <- celltypes
            } else if ("PBMC" == item) {
                ## set cell types for human PBMC data
                split_res <- strsplit(as.character(res$ID), "_")
                celltypes <- unlist(lapply(split_res, "[", 1))
                res$cell_type <- celltypes
            }
        } else {
            res$cell_type <- res$ID
        }

        result_list[[item]] <- res
    }
    result_list
}


#' Find most enriched cell types in bulk DE genes by Logistic Regression
#'
#' This is a function which takes marker genes from single-cell RNA-seq as
#' reference to calculate the enrichment of certain cell types in bulk DEG
#' analysis. We assume that bulk DEG is derived from certain cell-type specific
#' pattern.
#'
#' @name LRcellCore
#'
#' @param gene.p Named vector of gene-level pvalues from DEG analysis, i.e.
#' DESeq2, LIMMA
#'
#' @param marker.g List of Cell-type specific marker genes derived from
#' single-cell RNA-seq. The name of the list is cell-type or cluster name, the
#' values are marker genes vectors or numeric named vectors. LRcell provides
#' marker genes list in different human/mouse brains, but users could use their
#' own marker gene list as input.
#' default: NULL
#'
#' @param method Either `logistic regression` or `linear regression`. Logistic
#' regression equally treats cell-type specific marker genes, however, if
#' certain values could determine the importance of marker genes, linear
#' regression can be performed, default: LR.
#'
#' @param min.size Minimal size of a marker gene set, will impact the balance of
#' labels
#'
#' @param sig.cutoff Cutoff for input genes' pvalues, default: 0.05.
#'
#' @return A dataframe of LRcell statistics as described in \code{\link{LRcell}}.
#'
#' @import stats
#' @import utils
#'
#' @export
#'
#' @examples
#' data(mouse_FC_marker_genes)
#' data(example_gene_pvals)
#' res <- LRcellCore(example_gene_pvals, mouse_FC_marker_genes, method="LR")
LRcellCore <- function(gene.p,
                       marker.g,
                       method,
                       min.size = 5, sig.cutoff = 0.05) {

    if (typeof(marker.g) != "list")
        stop("Please make sure marker genes is a list with names indicating cell types or clusters.")

    if (method == "LiR") {
        # check whether marker genes are with statistics indicating significance
        if (all(lapply(marker.g, is_namedvector) == TRUE) == FALSE)
            stop("Please check each item in the list is a named vector with statistics for each gene.")
    }

    if (method == "LR") {
        # check whether marker genes are within each group
        if (all(lapply(marker.g, is_vector) == TRUE) == FALSE)
            stop("Please check each item is a vector containing gene symbols.")
    }

    # filter gene_pvals named vector
    gene.p <- gene.p[!is.na(gene.p)]
    gene.p[gene.p == 0] <- 10^(-15)
    gene_nlp <- -log(gene.p)

    # average duplicate genes
    if (length(names(gene_nlp)) != length(unique(names(gene_nlp)))) {
        dedup_gene_nlp <- tapply(gene_nlp, names(gene_nlp), mean)
        gene_nlp <- dedup_gene_nlp
    }
    sig_genes <- names(gene_nlp)[exp(-gene_nlp) < sig.cutoff]

    # filter single-cell marker gene list
    filtered_marker_genes <- marker.g[lapply(marker.g, length) >= min.size]

    # start analyze
    ids <- NA; coefs <- NA; pvals <- NA; genes_num <- NA; lead_genes <- NA
    idx <- 0;
    for (sub in names(filtered_marker_genes)) {
        marker_genes <- filtered_marker_genes[[sub]]
        if (method == "LiR") {
            # average duplicate marker genes
            dedup_marker_genes <- tapply(marker_genes, names(marker_genes), mean)
            marker_genes <- names(dedup_marker_genes)
        }
        matched_idx <- match(marker_genes, as.vector(names(gene_nlp), mode="character"))
        matched_idx <- matched_idx[!is.na(matched_idx)]
        matched_genes <- as.vector(names(gene_nlp), mode="character")[matched_idx]

        if (length(matched_genes) < min.size) next

        # start analyzing
        idx <- idx+1
        ids[idx] <- sub
        genes_num[idx] <- length(matched_genes)
        sig_DE_genes <- intersect(marker_genes, sig_genes)
        lead_genes[idx]<-paste(sig_DE_genes[order(sig_DE_genes)],collapse=", ")

        y <- rep(0, length(gene_nlp))
        if (method == "LR") {
            y[matched_idx] <- 1
            lr_df <- data.frame("y"=as.numeric(y), "x"=as.numeric(unname(gene_nlp)))
            # fit logistic regression
            glm.lr <- stats::glm(y ~ x,
                                   family = binomial(link="logit"),
                                   maxit = 100,
                                   lr_df)
            glm_res <- summary(glm.lr)
            coefs[idx] <- glm_res$coefficients["x","Estimate"]
            pvals[idx] <- glm_res$coefficients["x","Pr(>|z|)"]
        }

        if (method == "LiR") {
            matched_scores <- unname(dedup_marker_genes[matched_genes])
            y[matched_idx] <- as.numeric(matched_scores)
            lir_df <- data.frame("y"=y, "x"=as.numeric(unname(gene_nlp)))
            # fit linear regression
            glm.lir <- stats::glm(y ~ x,
                                family = gaussian(link="identity"),
                                maxit = 100,
                                lir_df)
            glm_res <- summary(glm.lir)
            coefs[idx] <- glm_res$coefficients["x","Estimate"]
            pvals[idx] <- glm_res$coefficients["x","Pr(>|t|)"]
        }
    }

    BH_fdr <- p.adjust(pvals,"BH")
    if (method == "LR") {
        # calculate the odds of marker gene being enriched when input pval
        # increase from 0.001 to 0.5
        odds_increase <- log(0.5)-log(0.001)
        odds_ratio <- exp(odds_increase*coefs)
        res <- data.frame("ID"=ids, "genes_num"=genes_num,
                        "coef"=coefs, "odds_ratio"=odds_ratio,
                        "p-value"=pvals, "FDR"=BH_fdr,
                        "lead_genes"=lead_genes)
    }

    if (method == "LiR") {
        res <- data.frame("ID"=ids, "genes_num"=genes_num,
                        "coef"=coefs, "p-value"=pvals, "FDR"=BH_fdr,
                        "lead_genes"=lead_genes)
    }
    res
}


#' Find most enriched cell types in bulk DE genes by Logistic Regression
#'
#' This is a function which takes marker genes from single-cell RNA-seq as
#' reference to calculate the enrichment of certain cell types in bulk DEG
#' analysis. This algorithm borrows from Marques et al, 2016 (https://science.sciencemag.org/content/352/6291/1326.long).
#'
#' @name LRcell_gene_enriched_scores
#'
#' @param expr Expression matrix with rows as genes and columns as cells, can be
#' an object of Matrix or dgCMatrix or a dataframe.
#'
#' @param annot Cell type annotation named vector with names as cell ids and
#' values as cell types.
#'
#' @param power The penalty on fraction of cells expressing the genes.
#'
#' @param parallel Whether to run it in parallel.
#'
#' @param n.cores How many cores to use in parallel mode.
#'
#' @return A numeric matrix with rows as genes and columns as cell types,
#' values are gene enrichment scores.
#'
#' @import BiocParallel
#' @export
LRcell_gene_enriched_scores <- function(expr,
                                     annot,
                                     power=1,
                                     parallel=TRUE,
                                     n.cores=4) {
    cat("Generate enrichment score for each gene..\n")
    gene_enriched_list <- NULL

    # check length
    if (ncol(expr) != length(annot))
        stop("Please check your provided cell type annotation is corresponding to the input expression dataframe.")
    # check corresponding cell names
    if(!all(sort(colnames(expr)) == sort(names(annot))))
        stop("Please check your provided cell type annotation is corresponding to the input expression dataframe.")


    if (parallel) { # do parallel
        BPPARAM <- BiocParallel::SnowParam(workers = n.cores,
                                           progressbar = TRUE,
                                           type = 'SOCK')

        gene_enriched_list <- BiocParallel::bplapply(rownames(expr),
                                                     enrich_posfrac_score,
                                                     expr,
                                                     annot,
                                                     power,
                                                     BPPARAM = BPPARAM)
    }
    else { # or just for-loop
        # set progress bar
        pb <- utils::txtProgressBar(max=nrow(expr), style=3)
        progress <- function(n) utils::setTxtProgressBar(pb, n)

        for (i in seq_len(nrow(expr))) {
            gene <- rownames(expr)[i]
            gene_enriched_list[[gene]] <- enrich_posfrac_score(gene,
                                                               expr,
                                                               annot,
                                                               power=power)
            progress(i)
        }
        close(pb)
    }

    cat("total gene number:", length(gene_enriched_list), '\n')
    enriched_genes <- do.call(rbind, gene_enriched_list)
    rownames(enriched_genes) <- rownames(expr)

    ## convert enriched gene scores dataframe into matrix
    numeric_enriched_genes <- apply(enriched_genes, c(1, 2), as.numeric)
    numeric_enriched_genes
}


#' Calculate enrichment scores for each cell type in a specific gene.
#'
#' This function takes a specific gene expression, cell type annotation and a
#' hyperparameter to calculate enrichment scores.
#'
#' @name enrich_posfrac_score
#'
#' @param gene Gene name from the expression matrix.
#'
#' @param expr Complete expression matrix with rows as genes and columns as cells.
#'
#' @param annot Cell type annotation named vector with names as cell ids and
#' values as cell types.
#'
#' @param power The penalty on fraction of cells expressing the genes
#'
#' @return Enrichment score list with cell type as names and enrichment score as
#' values.
enrich_posfrac_score <- function(gene, expr, annot, power=1) {
    gene.expr <- expr[gene, ]
    score_list <- list()
    for (celltype in unique(annot)) {
        cells <- names(annot[annot == celltype])
        cells_gene_expr <- gene.expr[cells]
        # enrichment in this cell type
        enrich <- (sum(cells_gene_expr)/length(cells_gene_expr)) / (sum(gene.expr)/length(gene.expr))
        # proportion of this gene express
        postfrac <- sum(cells_gene_expr > 0) / length(cells_gene_expr)
        score_list[[celltype]] <- enrich*(postfrac^power)
    }
    score_list
}

